### ISSUE TITLE
Critical memory issue since 1.0.21

### ISSUE THREAD

**Issue by MaximeGratens** [BuilderIO/builder#3242]
**Describe the bug**
Hello,
Since 1.0.21 and using the ini
**To Reproduce**
1. Install sdk react 2 with remix 
2. Use the initiate node sdk inside loader
3. Deploy in a production server and monitore

**Expected behavior**
The memory or cpu of the server will increase until it will crash

**Screenshots**
![image](https://github.com/BuilderIO/builder/assets/15386704/15ecdbcb-b377-4834-8685-e52b13fbd491)
* at 12 I deployed the 1.0.23 without the node setup

**Additional context**
We deploy the sdk 1.0.23 without the initiate node and you can see in the screen that the memory issue is fixed

**Comment by mikehuebner** [comment]
Seeing this as well on `1.0.23`, it just crashed our beanstalk application with a bunch of errors, removing the `initializeNodeRuntime` causes the below error. However, just removing the `initializeNodeRuntime` fixes the issue.

```
[Builder.io]:  Failed code evaluation: [Builder.io]: could not import `isolated-vm` module for safe script execution on Node server.      In certain Node environments, the SDK requires additional initialization steps. This can be achieved by   importing and calling `initializeNodeRuntime()` from "@builder.io/sdk-react/node/init". This must be done in   a server-only execution path within your application.   Please see the documentation for more information: https://builder.io/c/docs/integration-tips#enabling-data-bindings-in-node-environments    { code: 'state.posts' }
```

**Comment by MaximeGratens** [comment]
> Seeing this as well on `1.0.23`, it just crashed our beanstalk application with a bunch of errors, removing the `initializeNodeRuntime` causes the below error. However, just removing the `initializeNodeRuntime` fixes the issue.
> 
> ```
> [Builder.io]:  Failed code evaluation: [Builder.io]: could not import `isolated-vm` module for safe script execution on Node server.      In certain Node environments, the SDK requires additional initialization steps. This can be achieved by   importing and calling `initializeNodeRuntime()` from "@builder.io/sdk-react/node/init". This must be done in   a server-only execution path within your application.   Please see the documentation for more information: https://builder.io/c/docs/integration-tips#enabling-data-bindings-in-node-environments    { code: 'state.posts' }
> ```

If you remove the initializeNodeRuntime and use symbols, your site's performance will be degraded.

**Comment by samijaber** [comment]
We just published some perf improvements: reusing an Isolate VM context across all data binding evaluations. This should reduce memory usage during SSR.

Available at 1.0.25, does not require any change besides updating the version.

https://github.com/BuilderIO/builder/releases/tag/%40builder.io%2Fsdk-react%401.0.25

**Comment by samijaber** [comment]
@MaximeGratens If you can try this new version and let me know how it fares performance-wise, that would be great.

**Comment by MaximeGratens** [comment]
trying now ! (will edit this message)

**Comment by kaceycleveland** [comment]
As a work around for now @MaximeGratens , can pass an empty object to `initializeNodeRuntime`:

```
initializeNodeRuntime({})
```

**Comment by kevintruby** [comment]
Wow, what timing! I was going to open a similar issue _for the Vue SDK_, but now it might make more sense to comment here instead.

I've been delaying a major upgrade from the `@builder.io/sdk-vue` package's 10/2023 `0.7.0` release to the latest, because our E2E tests were experiencing periods of downtime when testing a staging site. Upon investigation it seemed the introduction of `isolated-vm` at version `0.7.1` or `0.7.2` (_could never get `0.7.1` to work_) alone resulted in a major memory leak.

It took some time to dig into what part of our content caused this, because some pages behaved fine. In our case, any version that included `isolated-vm` prior to this `1.0.25` update struggled with pages that included symbols _**that loop through a `List` input's contents and render images**_. Our use case was a symbol that could generate a row of CTAs for our e-commerce site.

I recreated a very narrow version of the issue with a quick Nuxt repo for SSR plus a lightweight Docker container that could be used to watch memory use over time, like so:

![Screenshot 2024-04-29 at 5 47 33 PM](https://github.com/BuilderIO/builder/assets/24785158/d6d8288d-0264-4947-b3d5-5ef381e357fa)

Will keep that around just in case, but I updated to `1.0.25` and the narrow reproduction no longer spikes. Applying this to my actual site's environment also appears to behave as expected. I'll know for sure after automated tests run on this build, but I feel confident after examining the local Docker container.

Figured this could be anecdotal evidence of the fix for other SSR environments 🤞

**Comment by MaximeGratens** [comment]
We are in production with the the 1.0.25 and the initializeNodeRuntime({});

We don't see for now crash or memory limit (we are doing our test on a multiple instance with 512mb of ram).

We still have some page really slow and this is the same as @kevintruby, it's when we are trying to loop.

You have this inside this issiue #3240

**Comment by MaximeGratens** [comment]
![image](https://github.com/BuilderIO/builder/assets/15386704/4fc4d977-987f-4798-832c-b6b48bd2968a)
Here is the last 6 hours with the the 1.0.25 in production.

As you can see we had a crash few minutes ago.

Also, the memory graph is completely jagged.

**Comment by samijaber** [comment]
@kevintruby Thanks for sharing all of these insights. By reusing Isolate contexts across all data-binding evaluations, the latest change should get rid of memory leaks. Do let me know if there are any persistent issues when you deploy your Vue SDK test to production.

@MaximeGratens Thanks for testing the latest version. It sounds like this issue may be specific to the builder feature of iterating over an array of dynamically fetched data? I will respond in your other issue.

**Comment by samijaber** [comment]
I am closing this issue as the specific memory leak caused by `isolated-vm` has been resolved by reusing a single IsolatedVM Context per render.

The unrelated performance issue tied to repeated blocks on data bindings can be tracked here https://github.com/BuilderIO/builder/issues/3240

**Comment by samijaber** [comment]
Additionally: I patched `initializeNodeRuntime` so that you do not need to provide an empty object as a parameter.

See `v1.0.26`: https://github.com/BuilderIO/builder/blob/main/packages/sdks/output/react/CHANGELOG.md#1026

**Comment by MaximeGratens** [comment]
I don't think the problem has been really solved.

2 grafs of 2 differents projects using builder and remix.

![image](https://github.com/BuilderIO/builder/assets/15386704/662c6b72-9dc8-4c70-98ea-b119c129f8aa)
![image](https://github.com/BuilderIO/builder/assets/15386704/40822d1c-b1bd-4c2a-a521-906d634faa71)

Both server are crashing after sometimes.

**Comment by khiemtong** [comment]
As another data point, I can confirm that this fix resolved our memory issues with `NextJS@14.2.3` and `@builder.io/sdk-react@1.0.25`

**Comment by samijaber** [comment]
Thanks folks. @MaximeGratens we are actively investigating the other sources of memory issues that you are facing and will respond to them in #3240
